README.md
# metl2011
C# WPF MeTL client

This project contains the source code for the MeTL 2011 client, a WPF application served through ClickOnce.

This requires a MeTL backend to be useful.

It was created at Monash University, in the eEducation center.


changes.txt
<logCollection date='{2010,7,21}'><message xml:lang='en' to='1099406@conference.deified.adm.monash.edu.au' from='tessn@deified.adm.monash.edu.au' type='groupchat' time='1279678271101'><image xmlns='monash:metl'><tag>{&quot;author&quot;:&quot;tessn&quot;,&quot;privacy&quot;:&quot;public&quot;,&quot;id&quot;:&quot;tessn:21/07/2010 12:16:48 PM:0&quot;,&quot;isBackground&quot;:false,&quot;zIndex&quot;:0}</tag><source>/Resource/1099406/background17.jpg</source><width>720</width><height>540</height><x>0</x><y>0</y><author>tessn</author><target>presentationSpace</target><privacy>public</privacy><slide>1099406</slide><identity>tessn:21/07/2010 12:16:48 PM:0</identity></image></message><message xml:lang='en' to='1099406@conference.deified.adm.monash.edu.au' from='tessn@deified.adm.monash.edu.au' type='groupchat' time='1279678271101'><image xmlns='monash:metl'><tag>{&quot;author&quot;:&quot;tessn&quot;,&quot;privacy&quot;:&quot;public&quot;,&quot;id&quot;:&quot;tessn:21/07/2010 12:16:48 PM:1&quot;,&quot;isBackground&quot;:false,&quot;zIndex&quot;:0}</tag><source>/Resource/1099406/background18.jpg</source><width>648</width><height>96.87504</height><x>30.12504</x><y>12</y><author>tessn</author><target>presentationSpace</target><privacy>public</privacy><slide>1099406</slide><identity>tessn:21/07/2010 12:16:48 PM:1</identity></image></message><message xml:lang='en' to='1099406@conference.deified.adm.monash.edu.au' from='tessn@deified.adm.monash.edu.au' type='groupchat' time='1279678271384'><image xmlns='monash:metl'><tag>{&quot;author&quot;:&quot;tessn&quot;,&quot;privacy&quot;:&quot;public&quot;,&quot;id&quot;:&quot;tessn:21/07/2010 12:16:48 PM:2&quot;,&quot;isBackground&quot;:false,&quot;zIndex&quot;:0}</tag><source>/Resource/1099406/background19.jpg</source><width>599</width><height>304.625031</height><x>30</x><y>145.25</y><author>tessn</author><target>presentationSpace</target><privacy>public</privacy><slide>1099406</slide><identity>tessn:21/07/2010 12:16:48 PM:2</identity></image></message><message xml:lang='en' to='1099406@conference.deified.adm.monash.edu.au' from='srsha7@deified.adm.monash.edu.au' type='groupchat' time='1279687834216'><ink xmlns='monash:metl'><checksum>6425</checksum><points>30.6 133.2 10 30.6 133.2 27 30.6 133.2 44 30.6 133.2 56 31 132.7 64 31 132.7 69 31 132.7 74 31 132.7 79 31.5 132.5 86 31.5 132.5 92 31.5 132.5 100 31.5 132.5 108 31.8 133 116 31.8 133 123 31.8 133 130 31.8 133 135 31.8 133 139 31.8 133 142 31.9 133.5 144 31.9 133.5 145 31.9 133.5 147 31.9 133.5 149 31.9 133.5 151 31.9 133.5 153 31.9 133.5 156 31.9 133.5 159 31.9 133.5 162 31.9 133.5 165 31.9 133.5 167 31.9 133.5 170 31.9 133.5 170 31.9 133.5 171 31.9 133.5 171 31.9 133.5 170 31.9 133.5 164 31.9 133.5 140 31.4 133.3 110 31.4 133.3 79 30.9 132.7 48</points><color>0 0 0 255</color><thickness>1</thickness><highlight>False</highlight><author>srsha7</author><target>presentationSpace</target><privacy>public</privacy><slide>1099406</slide></ink></message><message xml:lang='en' to='1099406@conference.deified.adm.monash.edu.au' from='srsha7@deified.adm.monash.edu.au' type='groupchat' time='1279687836687'><ink xmlns='monash:metl'><checksum>12607.4</checksum><points>33.9 135.9 2 34.4 136.1 14 34.4 136.1 25 34.4 136.1 35 34.4 136.1 45 34.4 136.1 57 34.4 136.1 70 34.4 136.1 83 34.4 136.1 96 34.4 136.1 107 34.4 136.1 117 34.6 135.6 125 34.6 135.6 133 34.6 135.6 138 34.6 135.6 142 34.5 135.2 145 34.5 135.2 149 34.5 135.2 152 34.5 135.2 155 34.5 135.2 157 34.2 134.6 160 34.2 134.6 162 33.9 134.1 165 33.9 134.1 167 33.2 133.6 170 33.2 133.6 171 32.6 133.3 173 32.6 133.3 174 32 133.4 176 32 133.4 177 31.2 133.8 178 30.7 133.9 179 30.3 134.2 180 29.8 134.5 181 29.4 134.8 181 29 135.2 181 28.6 135.6 182 28.3 136.2 181 28.1 136.7 181 28 137.3 182 28 137.9 182 28 138.5 182 28.1 139 182 28.3 139.4 182 28.3 139.4 182 29.1 140.1 183 29.7 140.2 183 30.5 140.3 183 31.4 140.4 183 32.4 140.5 183 33.3 140.6 183 34.3 140.8 183 35.3 141.2 183 36.1 141.8 183 36.8 142.6 183 37.3 143.7 183 37.5 144.9 184 37.5 146.3 185 37.5 147.8 186 37.2 149.4 187 36.5 150.9 188 35.7 152.5 190 34.7 153.9 191 33.5 155.1 192 32.2 156.1 194 30.8 156.8 194 29.2 157.2 194 27.7 157.4 191 26.1 156.9 179 25 156.4 146 23.8 155.5 112 23 154.4 77 22.2 153 43</points><color>0 0 0 255</color><thickness>1</thickness><highlight>False</highlight><author>srsha7</author><target>presentationSpace</target><privacy>public</privacy><slide>1099406</slide></ink></message><message xml:lang='en' to='1099406@conference.deified.adm.monash.edu.au' from='srsha7@deified.adm.monash.edu.au' type='groupchat' time='1279687837566'><ink xmlns='monash:metl'><checksum>9890.4</checksum><points>41.8 152.1 33 42 152.5 61 42 152.5 91 42.4 152.7 112 43 153.1 129 43 153.1 144 44.2 153.2 155 44.7 153 163 45.6 152.8 170 46.3 152.4 175 47.1 151.9 179 47.8 151.2 182 48.4 150.5 184 48.9 149.7 185 49.2 148.8 186 49.4 147.9 187 49.5 147.2 187 49.4 146.4 188 49.3 146 189 49.3 146 189 48.6 145.4 190 48.6 145.4 191 47.7 145.3 192 47.2 145.4 193 46.6 145.7 194 46 146 194 45.3 146.4 195 44.7 146.8 195 44 147.2 196 43.5 147.8 196 43.1 148.4 196 42.8 149.1 196 42.7 150 196 42.8 150.9 196 43.2 152 197 43.6 153.3 197 44.1 154.6 198 44.8 155.8 198 45.6 156.8 199 46.3 157.7 200 47.2 158.3 201 48 158.6 202 48.9 158.6 202 49.6 158.1 202 50.3 157.2 198 51.2 156.4 174 51.8 155 139 52.3 153.6 105 52.6 152 68 52.9 150.3 32</points><color>0 0 0 255</color><thickness>1</thickness><highlight>False</highlight><author>srsha7</author><target>presentationSpace</target><privacy>public</privacy><slide>1099406</slide></ink></message><message xml:lang='en' to='1099406@conference.deified.adm.monash.edu.au' from='srsha7@deified.adm.monash.edu.au' type='groupchat' time='1279687838129'><ink xmlns='monash:metl'><checksum>5422.1</checksum><points>59 135.8 49 59 135.8 81 59.2 136.4 113 59.2 136.8 146 59.2 136.8 161 59.1 137.8 174 59.1 138.4 182 59.2 139.4 187 59.3 140.4 190 59.5 141.8 193 59.8 143.2 196 60.1 144.9 198 60.3 146.7 201 60.7 148.

WebMeTL/app/controllers.scala
package controllers

import utils.Stemmer._
import play._
import play.mvc._
import play.libs._
import org.apache.commons.io._
import org.apache.commons.compress.archivers._
import org.apache.commons.compress.archivers.zip._
import java.io.{File}
import scala.xml.parsing._
import scala.collection.JavaConversions._
import org.json.simple._
import net.liftweb._
import net.liftweb.json._
import net.liftweb.json.JsonDSL._
import net.liftweb.json.JsonAST._
import collection.breakOut;
import MeTL._
import scala.math._

object Application extends Controller {
    val width = 200
    val height = 150
    val structure = "https://%s.adm.monash.edu.au:1188/Structure"
    val history = "https://%s.adm.monash.edu.au:1749/%s/%d/all.zip"
    val username = "exampleUsername"
    val password = "examplePassword"
    val TEMP_FILE = "all.zip"
    private def loadConversations(server:String) = {
        println("Loading conversations for "+server)
        val zipFuture = WS.url(structure.format(server)+"/all.zip").authenticate(username,password).get
        FileUtils.writeByteArrayToFile(new File(TEMP_FILE),IOUtils.toByteArray(zipFuture.getStream))
        val zipFile = new ZipFile(new File(TEMP_FILE))
        Map(zipFile.getEntries
            .map(any => any.asInstanceOf[ZipArchiveEntry])
            .filter(zae=>zae.getName.endsWith("details.xml"))
            .map(zae => IOUtils.toString(zipFile.getInputStream(zae)))
            .map(xmlString => xml.XML.loadString(xmlString))
            .map(detail => ((detail \ "jid").text.toInt -> detail)).toList:_*)
    }
    private def mapRandom(history:Map[Int,xml.NodeSeq], max:Int = 10) = history.map(kv => (kv._1 -> (Math.random * max).intValue.toString))
    def index = conversations("deified")
    def conversations(server:String) = {
        val serverName = server
        val authorJson = authorSummaries(server).toString()
        Template(serverName,authorJson)
    }
    def conversation(server:String,jid:Int)={
        val slides = loadConversations(server)(jid) \\ "slide"
        val messages = slides
            .filter(s=> (s \\ "type").text == "SLIDE")
            .map(s=>slide(server,(s \\ "id").text.toInt))
            .flatten
        val slideOrders = Map(slides.map(node=>((node \\ "id").text.toInt -> (node \\ "index").text.toInt)).toList:_*)
        val authors = Map(messages.map(m=>m.author).distinct.map(a=>(a->(Math.random*5).toInt)):_*)
        val relativizedMessages = messages.map(m=>Message(m.name,m.timestamp,slideOrders(m.slide), m.author, authors(m.author)))
        //val clump = SimpleClump(messages)
        //val clump = MessageReductor.clumpSlides(messages)
        val clump = MessageReductor.clump(relativizedMessages, true)
        pretty(render(clump.toJsonFull))
    }
    private def slideXmppMessages(server:String,jid:Int)={
        val uri = history.format(server,stem(jid.toString),jid)
        println("Retrieving "+uri)
        val zipFuture = WS.url(uri).authenticate(username,password).get
        FileUtils.writeByteArrayToFile(new File(TEMP_FILE),IOUtils.toByteArray(zipFuture.getStream))
        val zipFile = new ZipFile(new File(TEMP_FILE))
        zipFile.getEntries
            .map(any => any.asInstanceOf[ZipArchiveEntry])
            .filter(zae=>zae.getName.endsWith(".xml"))
            .map(zae => IOUtils.toString(zipFile.getInputStream(zae))+"</logCollection>")
    }
    private def slide(server:String,jid:Int)={
        slideXmppMessages(server,jid)
            .map(detail => xml.XML.loadString(detail))
            .foldLeft(List.empty[Seq[Message]])((acc,item)=>{
                (item \\ "message").flatMap(message=>{
                    val timestamp = (message \ "@time").text.toLong
                    List("image","ink","text")
                        .flatMap(nodeName=>{
                            (message \\ nodeName).map(node=>{
                                val author = (node \ "author").text
                                Message(nodeName,timestamp,jid,author,0)
                           })
                        })
                }) :: acc
        }).flatten
    }
    def quizzes(server:String,jid:Int)={
        val qs = slideXmppMessages(server,jid)
            .map(d=>xml.XML.loadString(d))
            .foldLeft(List.empty[Quiz])((acc,item)=>{
                (item \\ "message").flatMap(message=>{
                    val timestamp = (message \ "@time").text.toLong
                    (message \ "quiz").map(quiz=>{
                        println(quiz)
                        val title = (quiz \ "title").text
                        val id = (quiz \ "id").text.toLong
                        val conversation = jid
                        val author = (quiz \ "author").text
                        val options = (quiz \ "quizOption").map(QuizOption.parse(_))
                        Quiz(title,id,timestamp,conversation,author,options)
                    })
                }).toList ::: acc
            })        
        pretty(render(JArray(qs.map(_.toJson))))
    }
    def snapshot(width:Int=640, height:Int=320,server:String,slide:Int) ={
        println("Snapshotting "+slide)
        val image = viewModels.Snapshot.png(server,width,height,slideXmppMessages(server,slide).toList)
        javax.imageio.ImageIO.write(image, "png", response.out)
        request.contentType = "image/png"
    }
    private def authorFrequencies(xs:Map[Int,xml.NodeSeq]) = {
        val authors = xs.map(kv => (kv._2 \ "author").text)
        authors.groupBy(identity).mapValues(_.size)
    }
    private def elem(name:String, children:xml.Node*) = xml.Elem(null, name ,xml.Null,xml.TopScope, children:_*); 
    private def node(name:String, parent:xml.NodeSeq) = elem(name, xml.Text((parent \ name).text))
    private def authorSummary(kv:Pair[String,Int], xs:Map[Int,xml.NodeSeq]) = {
        val contentWeight = mapRandom(xs, 1000)
        val authorsPerConversation = mapRandom(xs, 20)
        val author = kv._1
        val freq = kv._2
        val count = elem("conversationCount", xml.Text(freq.toString))
        val conversationsBody = xs.filter(kv=>(kv._2 \ "author").text == author).map(
            kv=>{
                val node = kv._2
                val conversationContentWeight = elem("contentVolume", xml.Text(contentWeight((node \ "jid").text.toInt)))
                val authorCount = elem("authorCount", xml.Text(authorsPerConversation((node \ "jid").text.toInt)))
                val slideCount = elem("slideCount", xml.Text((node \ "slide").length.toString))
                <listing>
                    {List("title","jid").map(name=>this.node(name, node)) ::: List(conversationContentWeight, authorCount, slideCount)}
                </listing>;
            }).toList
        val conversations = <conversations>{conversationsBody}</conversations>;
        elem(author, List(count, conversations):_*)
    }
    private def authorSummaries(server:String)={
        val xs = loadConversations(server)
        val authorXmlList = authorFrequencies(xs).map(kv=>authorSummary(kv, xs))
        val authors = 
        <conversationSummaries>
            {authorXmlList}
        </conversationSummaries>;
        val authorsJson = json.Xml.toJson(authors).transform {
            case JField("listing", x: JObject) => JField("listing", JArray(x :: Nil))
            case other => other
        }
        pretty(render(authorsJson))
    }
}
